src/rpc/budget.cpp

```diff
@@ -371,16 +373,17 @@ static mnKeyList getDMNKeys(CWallet* const pwallet, const Optional<std::string>&
         }
     }
 
-    LOCK(pwallet->cs_wallet);
-    EnsureWalletIsUnlocked(pwallet);
-
     mnKeyList mnKeys;
     mnList.ForEachMN(true, [&](const CDeterministicMNCPtr& dmn) {
         bool filtered = mnFilter && dmn->proTxHash == mnFilter->proTxHash;
         if (!mnFilter || filtered) {
             const CKeyID& mnKeyID = fFinal ? dmn->pdmnState->keyIDOperator : dmn->pdmnState->keyIDVoting;
             CKey mnKey;
-            if (pwallet->GetKey(mnKeyID, mnKey)) {
+            {
+                LOCK(pwallet->cs_wallet);
+                pwallet->GetKey(mnKeyID, mnKey);
+            }
+            if (mnKey.IsValid()) {
```

Comment:
better to improve this in another focused PR and remove the for-loop entirely. The node/wallet can know and maintain an indicator and/or a cached list of the enabled DMNs for which it can vote for.

Reply:
Sure, but in the meantime we can (and imo should) keep it as it was: LOCK outside of the loop, and return value of `GetKey` checked.